compiler/qsc_data_structures/src/language_features.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use clap::ValueEnum;
+use serde::Deserialize;
+use std::collections::BTreeSet;
+#[derive(Deserialize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Copy)]
+#[serde(rename_all = "kebab-case")]
+pub enum LanguageFeature {
+    /// This language feature enables experimental syntax that will likely be stabilized in the next major version.
+    /// It may include removals of outdated syntax and introductions of new syntax.
+    V2PreviewSyntax,
+}
+
+impl LanguageFeature {
+    pub fn try_parse(s: &str) -> Result<Self, UnrecognizedLanguageFeature> {
+        match s {
+            "v2-preview-syntax" => Ok(LanguageFeature::V2PreviewSyntax),
+            _ => Err(UnrecognizedLanguageFeature::msg(format!(
+                "Unrecognized language feature: {s}"
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Default)]
+pub struct LanguageFeatures(BTreeSet<LanguageFeature>);
+pub type LanguageFeatureIncompatibility = miette::ErrReport;
+pub type UnrecognizedLanguageFeature = miette::ErrReport;
+
+impl LanguageFeatures {
+    /// Checks that the current set of language features is compatible and well-formed.
+    /// Returns a descriptive error message if not.
+    pub fn check_compatibility(&self) -> Result<(), LanguageFeatureIncompatibility> {
```

Comment:
Since this method is currently not really doing anything, I think it would be better to not introduce it preemptively. If we add features in the future that are not compatible with each other, we can add a compatibility check then.